pkg/artifactory/datasource_artifactory_file.go

```diff
@@ -0,0 +1,167 @@
+package artifactory
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"github.com/atlassian/go-artifactory/v2/artifactory"
+	"github.com/atlassian/go-artifactory/v2/artifactory/v1"
+	"github.com/hashicorp/terraform/helper/schema"
+	"io"
+	"os"
+)
+
+func dataSourceArtifactoryFile() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceFileRead,
+
+		Schema: map[string]*schema.Schema{
+			"repository": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"path": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"created": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"created_by": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"last_modified": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"modified_by": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"last_updated": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"download_uri": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"mimetype": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"size": {
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+			"md5": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"sha1": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"sha256": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"output_path": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"force_overwrite": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+		},
+	}
+}
+
+func dataSourceFileRead(d *schema.ResourceData, m interface{}) error {
+	c := m.(*artifactory.Artifactory)
+
+	repository := d.Get("repository").(string)
+	path := d.Get("path").(string)
+	outputPath := d.Get("output_path").(string)
+	forceOverwrite := d.Get("force_overwrite").(bool)
+
+	fileInfo, _, err := c.V1.Artifacts.FileInfo(context.Background(), repository, path)
+	if err != nil {
+		return err
+	}
+
+	skip, err := SkipDownload(fileInfo, outputPath)
```

Comment:
Now that `output_path` is required, do we still need this? Maybe the verification of the SHAs can be here instead

Reply:
I've tried to refactor that part, however I believe it's not safe to verify the checksum only. There are multiple reasons why the checksum verification could fail:

- file doesn't exist
- any problems with the hash library
- io issues

However we only want to proceed with the download in the first case. Having that said, we'll still need to check whether the file exists or not.